### ISSUE TITLE
target controller as constructor parameter

### ISSUE THREAD

**Issue by sockeqwe** [bluelinelabs/Conductor#5]
I'm not 100% sure, but I think that the current way `Controller.setTargetController()` could cause problems because one could change the target controller afterwards. I think it would be better to set the target controller as constructor parameter to avoid such issues.

**Comment by sockeqwe** [comment]
Furthermore, I'm not sure how easy it is to implement, but usually a controller doesn't interact with the target controller "directly" but rather program against an interface as "listener". I think this is already an awkward design choice in android fragment (and iOS Controller too). Is there a way to introduce a special subclass `TargetController extends Controller` with a generic type for "listener". Or maybe this could be striped out of Controller somehow into `RouterTransaction` ...

I know it's really hard to implement something like this (screen orientation changes, process death etc.) but maybe the `Router` and `Backstack` can be smart enough somehow to detect such dependencies and ensure through `RouterTransaction` that the  "targetController / listener" is setup properly.

Just an idea, since we are already talking about that topic, then it might be considerable to think about alternative approach as well

**Comment by EricKuck** [comment]
I love the idea of adding generic listeners instead of target controllers. I originally explored this idea quite a bit, but couldn't figure out a way to reliably restore arbitrary objects in the case of Android killing the process. The nice thing about having a target controller is that Conductor will always be aware of them and can easily restore states when needed. If an arbitrary object is used, there's really no way (that I know of) to ensure that it'll be recreated, or even if it is recreated, that a reference can be obtained.

If you (or anyone else) has some insight into a way this could be done, I'd very much like to look into it. I've never liked the idea of setting a generic target either.

**Comment by EricKuck** [comment]
Oh, and regarding taking a target as a constructor param, are you suggesting that the `setTargetController` method be removed altogether? Or that another constructor be added which takes a target? I like the idea of adding that constructor option, as 99% of the time you'd want to set a target controller, it would be on instantiation.

**Comment by sockeqwe** [comment]
Sorry, I rushed ... 
My initial intention was to ask if there is a reason why one could set target controller afterwards?
Because the current implementation `Controller.setTargetController()` would allow to change the target controller at runtime, which might cause problems (or doesn't it?). Therefore, I was suggesting to set the target controller only by constructor. So the default constructor should be `protected Controller(Bundle args, Controller targetController)` and remove `setTargetController()` (not sure if possible because of `Controller.newInstance()`, but then at least remove public modifier).

> I originally explored this idea quite a bit, but couldn't figure out a way to reliably restore arbitrary objects in the case of Android killing the process.

I'm stuck on that too, but I will think about it ... Also I only see a few cases where I have to / want to communicate with a parent/target controller. Usually both controller should share the same model and controller should be notified on model ...

**Comment by EricKuck** [comment]
Setting the target controller at runtime is actually fine, unless the dev decides to set it as the host Activity is being destroyed.

The reason I'm hesitant to promote target controller to being in the primary constructor is exactly what you mentioned - there aren't that many cases where controllers should be communicating with each other directly, as long as you use a non-sucky architecture.

The current implementation would still allow people to set targets in the constructor, it would just be like this:

```
public <T extends Controller & MyListenerInterface> MyController(Bundle args, T target) {
    super(args);
    setTargetController(target);
}
```

instead of like this:

```
public <T extends Controller & MyListenerInterface> MyController(Bundle args, T target) {
    super(args, target);
}
```

So, knowing that setting the target controller at runtime shouldn't cause issues, are you still thinking this change should be made? I'm kind of on the fence about it.

**Comment by sockeqwe** [comment]
I'm fine with that!
I was just saying that the current API is from my point of view a little bit verbose. I.e. you could do something like this:
```java
Controller a = new Controller();

// Imagine you want to display a Dialog where the user can click YES or NO
Controller dialogController = new DialogController();

dialogController.setTargetController(a);

...
// Somewhere later in your code
Controller b = new Controller();
dialogController.setTargetController(b);
```

I was just saying that this kind of thing could cause problems, because now `Controller a` no longer is a callback. Of course it's the programmers fault, but with an API that allows only to set the target Controller once (like targetController as constructor parameter) such kind of things could be avoided.

However, if you are fine with the current API feel free to close this issue. Propably it's just nitpicking, sorry!

**Comment by EricKuck** [comment]
Ah, I hadn't thought of that. I totally get where you're coming from now. I agree that this is a good change given your new example. 

I suppose maybe it would be best to make this an optional constructor param? So we'd end up with `Controller()` for simple controllers that need no inputs or outputs, `Controller(Bundle args)` for controllers with  inputs but no outputs, and `Controller(Bundle args, Controller target)` for controllers with both? Or is this too confusing?

**Comment by sockeqwe** [comment]
Yeah it's not super nice to have so many constructors but it is reasonable to do that in constructor. Alternatively, target controller as part of the transaction? Also not that obvious I guess.

**Comment by EricKuck** [comment]
After going down the multiple constructor route for a while, it ultimately felt a bit too weird to have so many options for constructors. I believe I've addressed this issue a different way though - by making `setTargetController` not be final anymore. Controllers are now free to reject `setTargetController()` calls if they want to. The [target controller example](https://github.com/bluelinelabs/Conductor/blob/8a70c09fea06765b6e17a9ded4b9a87edc3e1741/demo/src/main/java/com/bluelinelabs/conductor/demo/controllers/TargetTitleEntryController.java) has been updated to show how this can be done. This example throws an exception, but they could also silently no-op the call for a less destructive response.

Hopefully that addresses your concerns. Feel free to reopen if not.

**Comment by sockeqwe** [comment]
I agree, another constructor parameter is too weird.

I'm ok with your solution, thought it can still lead to errors if not override correctly.

Alternatively, what about somethink like this:

```java
class Controller {
  ...
  public final setTargetController(Controller targetController){
       if (mTargetInstanceId != null) {
            throw new RuntimeException("Target Controller already set. You can set the target controller only once");
       }

        // Set the target controller
        mTargetInstanceId = target != null ? target.getInstanceId() : null;
   }

}

**Comment by EricKuck** [comment]
Great idea. This change was made in b5d0e46.
